File: go/internal/e2e/system_message_sections_e2e_test.go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

package e2e

import (
	"strings"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

func TestSystemMessageSectionsE2E(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	t.Run("should_use_replaced_identity_section_in_response", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			SystemMessage: &copilot.SystemMessageConfig{
				Mode: "customize",
				Sections: map[string]copilot.SectionOverride{
					"identity": {
						Action:  copilot.SectionActionReplace,
						Content: "You are a helpful gardening assistant called Botanica. You only answer questions about plants and gardening.",
					},
				},
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		response, err := session.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Who are you?",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}
		if response == nil {
			t.Fatal("Expected a response from the assistant")
		}

		ad, ok := response.Data.(*copilot.AssistantMessageData)
		if !ok {
			t.Fatalf("Expected AssistantMessageData, got %T", response.Data)
		}
		content := strings.ToLower(ad.Content)
		if !strings.Contains(content, "botanica") && !strings.Contains(content, "garden") && !strings.Contains(content, "plant") {
			t.Errorf("Expected response to reflect the replaced identity section, but got: %s", ad.Content)
		}
	})
}

File: nodejs/test/e2e/system_message_sections.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { describe, expect, it } from "vitest";
import { approveAll } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

describe("System message sections", async () => {
    const { copilotClient: client } = await createSdkTestContext();

    it("should_use_replaced_identity_section_in_response", async () => {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            systemMessage: {
                mode: "customize",
                sections: {
                    identity: {
                        action: "replace",
                        content:
                            "You are a helpful gardening assistant called Botanica. You only answer questions about plants and gardening.",
                    },
                },
            },
        });

        const response = await session.sendAndWait({ prompt: "Who are you?" });

        expect(response).not.toBeNull();
        const content = response!.data.content.toLowerCase();
        expect(
            content.includes("botanica") || content.includes("garden") || content.includes("plant"),
            `Expected response to reflect the replaced identity section, but got: ${response!.data.content}`
        ).toBe(true);

        await session.disconnect();
    });
});
